### PR TITLE
[CNFs] Test VPP pipeline service chain use case with the CNF Conformance Test Suite

### PR DESCRIPTION
**Pull request by wavell**
# [CNFs] Test VPP pipeline service chain use case  with the CNF Conformance Test Suite

## Description

Test use case from CNF Testbed.  Make minimal changes needed to deploy and run test.  Identify updates needed in test code.

## Issues:
Fixes: #143

## How has this been tested:
 - [ ] Covered by existing integration testing
 - [x] Added integration testing to cover
 - [x] Verified all A/C passes
     * [x] develop
     * [ ] master
     * [ ] tag/other branch
 - [x] Test environment
    * [x] Shared Packet K8s cluster
    * [ ] New Packet K8s cluster
    * [ ] Kind cluster
 - [ ] Have not tested

## Types of changes:
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
 - [ ] Documentation update

## Checklist:
**Documentation**
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] No updates required.

**Code Review**
- [ ] Does the test handle fatal exceptions, ie. rescue block

**Issue**
- [x] Tasks in issue are checked off

### CODE REVIEW

File: example-cnfs/ip-forwarder/csp/values.yaml [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    mac: ['52:54:01:00:02:aa','52:54:01:00:00:bb']
+    owner: ['slave','slave']
+    queues: 1
+    subnet: ['172.16.31.11/24','172.16.20.11/24']
+    remip: ['172.16.31.10','172.16.20.101']
+    remmac: ['52:54:01:00:01:bb','{{ index .Values.nfvbench_macs 1 }}']
+
+  5:
+    main_core: 38
+    worker_cores: '22,50'
```
Comment (taylor) [inline example-cnfs/ip-forwarder/csp/values.yaml:68]: This is hard coded cpu core pinning

File: example-cnfs/ip-forwarder/cnf-conformance.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+---
+helm_directory: csp
```
Comment (taylor) [inline example-cnfs/ip-forwarder/cnf-conformance.yml:2]: Csp is the name of the use case.  This is fine for a PoC

File: example-cnfs/ip-forwarder/csp/Chart.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
+appVersion: "1.0"
+description: 3c2n Pipeline Service Chain
```
Comment (taylor) [inline example-cnfs/ip-forwarder/csp/Chart.yaml:3]: Use case description vs CNF description

File: example-cnfs/ip-forwarder/csp/templates/configmap.yaml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    create memif socket id 2 filename {{ $.Values.volumeMounts.vpp_sockets.mountPath }}/{{ index .sock 1 }}.sock
+    create interface memif id {{ index .memid 0 }} socket-id 1 hw-addr {{ index .mac 0 }} {{ index .owner 0 }} rx-queues {{ .queues }} tx-queues {{ .queues }}
+    create interface memif id {{ index .memid 1 }} socket-id 2 hw-addr {{ index .mac 1 }} {{ index .owner 1 }} rx-queues {{ .queues }} tx-queues {{ .queues }}
+    set int ip addr memif1/{{ index .memid 0 }} {{ index .subnet 0 }}
+    set int ip addr memif2/{{ index .memid 1 }} {{ index .subnet 1 }}
+    set int state memif1/{{ index .memid 0 }} up
+    set int state memif2/{{ index .memid 1 }} up
+    set ip arp static memif1/{{ index .memid 0 }} {{ index .remip 0 }} {{ tpl ( index .remmac 0 ) $ }}
+    set ip arp static memif2/{{ index .memid 1 }} {{ index .remip 1 }} {{ tpl ( index .remmac 1 ) $ }}
+    ip route add 172.16.64.0/18 via {{ index .remip 0 }}
```
Comment (taylor) [inline example-cnfs/ip-forwarder/csp/templates/configmap.yaml:23]: Hard-coded IPs. Works for a PoC

File: example-cnfs/ip-forwarder/csp/templates/deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            items:
+            {{ range $k, $v := .Values.cnf }}
+            - key: startup_{{$k}}.conf
+              path: startup_{{$k}}.conf
+            - key: setup_{{$k}}.gate
+              path: setup_{{$k}}.gate
+            {{ end }}
+        - name: vpp-sockets
+          hostPath:
+            path: /etc/vpp/sockets/
```
Comment (taylor) [inline example-cnfs/ip-forwarder/csp/templates/deployment.yaml:54]: Volume mounts to host.  Works for PoC

**Comment by michaelspedersen** [comment]
I am in the middle of creating a simplified version of this CNF, which I think will be a better fit for use with the Conformance testing.
The one included here requires memif interfaces created in the host (host vSwitch), and the configuration file for VPP is made to support an external traffic generator (NFVbench) with additional CNFs also being present in the chain.

**Comment by taylor** [comment]
Ticket updated to match with items needed to adjust https://github.com/cncf/cnf-conformance/issues/143#issuecomment-616767412

**Pull request merged**